### PR TITLE
Add RequiresAccessedThrough test to NativeAOT

### PR DESCRIPTION
**Pull request by tlakollo**
Modify the DisplayNameHelpers to match linker way of printing generics
Add the RequiresAccessedThrough test from the linker and add the necessary
NativeAOT annotations
Add a reference to NativeAOT 'new' constraint issue

### CODE REVIEW

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs [hunk 1]
```diff
@@ -116,7 +117,58 @@ public static string GetDisplayName(this EventPseudoDesc @event)
...
-            return Formatter.Instance.FormatName(type, FormatOptions.NamespaceQualify);
+            var sb = new StringBuilder();
+            if (type == null)
+                return sb.ToString();
+
+            Stack<TypeDesc> genericArguments = null;
+            switch (type)
+            {
+                // TODO: Linker handles here ArrayType and GenericInstanceType
+                default:
```
Comment (vitek-karas) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:128]: This looks really weird - should we remove the switch for now?

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs [hunk 2]
```diff
@@ -116,7 +117,58 @@ public static string GetDisplayName(this EventPseudoDesc @event)
...
+                    string simpleName = Formatter.Instance.FormatName(type, FormatOptions.NamespaceQualify);
```
Comment (vitek-karas) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:129]: It's not "simple" name - it's namespace qualified name, right?

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs [hunk 3]
```diff
@@ -116,7 +117,58 @@ public static string GetDisplayName(this EventPseudoDesc @event)
...
+                    if (type.HasInstantiation)
+                    {
+                        int genericParametersCount = type.Instantiation.Length;
+                        int declaringTypeGenericParametersCount = 0;
+                        if (type is DefType defType)
+                            declaringTypeGenericParametersCount = defType.ContainingType?.Instantiation.Length ?? 0;
+
+                        if (genericParametersCount > declaringTypeGenericParametersCount)
+                        {
+                            if (genericArguments?.Count > 0)
```
Comment (vitek-karas) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:139]: As far as I can tell the `genericArguments` is never filled with any data, or is it?

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs [hunk 4]
```diff
@@ -116,7 +117,58 @@ public static string GetDisplayName(this EventPseudoDesc @event)
...
+                                PrependGenericArguments(genericArguments, genericParametersCount - declaringTypeGenericParametersCount, sb);
+                            else
+                                PrependGenericParameters(type.Instantiation, sb);
+
+                            int explicitArityIndex = simpleName.IndexOf('`');
+                            simpleName = explicitArityIndex != -1 ? simpleName.Substring(0, explicitArityIndex) : simpleName;
```
Comment (vitek-karas) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:145]: What happens for nested types? I think there will be backticks on multiple segments potentially.

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs [hunk 5]
```diff
@@ -116,7 +117,58 @@ public static string GetDisplayName(this EventPseudoDesc @event)
 
         public static string GetDisplayName(this TypeDesc type)
         {
-            return Formatter.Instance.FormatName(type, FormatOptions.NamespaceQualify);
+            var sb = new StringBuilder();
+            if (type == null)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:121]: Asking for a display name of null sounds like a bug. We're going to print a useless message to the user.

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs (hunk 2 above)
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:129]: Should we fix the formatter instead? Or is this how illink is doing it? (Trying to make sense of a string?)

I honestly wish we stopped trying to C#ify our names. Reflection doesn't do it either. It's fine.

Comment (tlakollo) [reply src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs]: Reverted the changes in the DisplayNameHelpers in favor of the solution that is currently in the main branch that modifies the result checker instead.

File: src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs [hunk 6]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Collections.Generic;
```
Comment (vitek-karas) [inline src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs:5]: Nit - left over from the changed to this - file - just revert the file to the original.

File: src/coreclr/tools/aot/Mono.Linker.Tests.Cases/RequiresCapability/RequiresAccessedThrough.cs [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+				[RequiresDynamicCode ("Message for --PropertyWithLdToken.get--")]
+				get {
+					return false;
+				}
+			}
+
+			[ExpectedWarning ("IL2026", "--PropertyWithLdToken.get--")]
+			[ExpectedWarning ("IL2026", "--PropertyWithLdToken.get--", ProducedBy = ProducedBy.Trimmer)]
+			[ExpectedWarning ("IL3002", "--PropertyWithLdToken.get--", ProducedBy = ProducedBy.Analyzer)]
+			[ExpectedWarning ("IL3050", "--PropertyWithLdToken.get--", ProducedBy = ProducedBy.Analyzer)]
```
Comment (vitek-karas) [inline src/coreclr/tools/aot/Mono.Linker.Tests.Cases/RequiresCapability/RequiresAccessedThrough.cs:228]: How come this is not produced by NativeAOT? - we should add a bug.

**Review by vitek-karas — APPROVED** [review]

**Pull request merged**